3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/apirequest/Input.scala

```diff
@@ -0,0 +1,236 @@
+
+// Scala
+import scala.util.control.NonFatal
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+// Json4s
+import org.json4s._
+import org.json4s.jackson.compactJson
+
+// This project
+import outputs.EnrichedEvent
+import utils.JsonPath._
+
+
+/**
+ * Container for key with one (and only one) of possible input sources
+ * Basically, represents a key for future template context and way to get value
+ * out of [[EnrichedEvent]], custom context, derived event or unstruct event.
+ *
+ * @param key extracted key
+ * @param pojo optional pojo source to take stright from `EnrichedEvent`
+ * @param json optional JSON source to take from context or unstruct event
+ */
+case class Input(key: String, pojo: Option[PojoInput], json: Option[JsonInput]) {
+  import Input._
+
+  // Constructor validation for mapping JSON to `Input` instance
+  (pojo, json) match {
+    case (None, None)       => { throw new MappingException("Input must represent either json OR pojo, none present") }
+    case (Some(_), Some(_)) => { throw new MappingException("Input must represent either json OR pojo, both present") }
+    case _ =>
+  }
+
+  /**
+   * Get key-value pair input from specific `event` for composing
+   *
+   * @param event currently enriching event
+   * @return template context with empty or with single element this particular input
+   */
+  def getFromEvent(event: EnrichedEvent): TemplateContext = pojo match {
+    case Some(pojoInput) => {
+      try {
+        val method = event.getClass.getMethod(pojoInput.field)
+        val value = Option(method.invoke(event)).map(_.toString)
+        value.map(v => Map(key -> Tags.LastVal(v))).successNel
+      } catch {
+        case NonFatal(err) => s"Error accessing POJO input field [$key]: [$err]".failureNel
+      }
+    }
+    case None => emptyTemplateContext
+  }
+
+  /**
+   * Get value out of list of JSON contexts
+   *
+   * @param derived list of self-describing JObjects representing derived contexts
+   * @param custom list of self-describing JObjects representing custom contexts
+   * @param unstruct optional self-describing JObject representing unstruct event
+   * @return template context with empty or with single element this particular input
+   */
+  def getFromJson(derived: List[JObject], custom: List[JObject], unstruct: Option[JObject]): TemplateContext =
+    json match {
+      case Some(jsonInput) => {
+        val validatedJson = jsonInput.field match {
+          case "derived_contexts" => getBySchemaCriterion(derived, jsonInput.schemaCriterion).successNel
+          case "contexts" => getBySchemaCriterion(custom, jsonInput.schemaCriterion).successNel
+          case "unstruct_event" => getBySchemaCriterion(unstruct.toList, jsonInput.schemaCriterion).successNel
+          case other => s"Error: wrong field [$other] passed to Input.getFromJson. Should be one of: derived_contexts, contexts, unstruct_event".failureNel
+        }
+
+        val validatedJsonPath = compileQuery(jsonInput.jsonPath)
```

Comment:
Yes! I think we can even short-circuit enrichment process if invalid JSON Path was given, instead of getting full enriched/bad bucket.